Bringing changes from sidekiq-alive-next

arturictus opened the pull request:
This change adds the following:

* sidekiq 7 support
* Updates to CI to test using multiple sidekiq versions
* Custom callback to run on shitdown (can be used to clean the queues)
* Some minor tweaks and refactors

arturictus commented:
Hi @andrcuns, 
I'm bringing all the goodies that I found in you master branch.
I let you as reviewer in the GitHub actions, we can talk about it if you are interested into getting access to this repo.

andrcuns commented:
> Hi @andrcuns, I'm bringing all the goodies that I found in your master branch. I let you as reviewer in the GitHub actions, we can talk about it if you are interested into getting access to this repo.

@arturictus thanks, that would be cool to release these under this gem for everyone else to use. Though since I mostly did the changes for my own usecase, some of them probably should be changed a bit, like:

- Changes for sidekiq >= 7 are not backwards compatible and it's probably best to hold off the rack upgrade to v3 as it's not actually currently compatible with latest Rails version.

I can try and help out in some capacity with maintaining this project at some capacity given I extensively use it in on of my other projects and it has been very helpful.

arturictus commented:
Hi @andrcuns,
In my case I'm not using much this gem any more and I'm missing the feedback of using it in production.
Would be nice to have some help from someone that is using it live in production.
do you want me to give you access to the repo?

andrcuns commented:
@arturictus Sure! Then I can try and update this branch to be a bit more backwards compatible sometime next week.

arturictus commented:
> ry and update this branch to be a bit more backwards compatible sometime next week.



> @arturictus Sure! Then I can try and update this branch to be a bit more backwards compatible sometime next week.

Cool!!, I give you access :)

arturictus commented on .github/workflows/test.yml:
```diff
@@ -1,19 +1,38 @@
-        ruby-version: ["3.1", "3.0", "2.7", "2.6", "2.5"]
-      # Service containers to run with `runner-job`
+        ruby-version: ["3.1", "3.0", "2.7"]
+        sidekiq-version: ["~> 5.2", "~> 6.5", "~> 7"]
```
This is great!!, great job!! 👍

arturictus commented on sidekiq_alive.gemspec:
```diff
@@ -1,40 +1,59 @@
-require 'sidekiq_alive/version'
+
+# allow testing multiple sidekiq version
+sidekiq_version_range = ENV["SIDEKIQ_VERSION_RANGE"] || "< 8"
```
I think this is better done in the gemfile.
here we just do:

```ruby
spec.add_dependency("sidekiq")
```

arturictus commented:
Hey @andrcuns, this looks great!
I added the secret: `GEM_HOST_API_KEY`
What is the `RELEASE_SSH_KEY`?
do we need it to publish the gem or it's for the release commit?

arturictus commented on .github/release.yml:
```diff
@@ -0,0 +1,23 @@
+changelog:
```
How do you use this semantic versioning?

arturictus replied:
do we need to use this: https://www.npmjs.com/package/semantic-release ?

andrcuns replied:
Nop, it's a github native changelog generation thing which generates entries based on labels in pull requests:
<img width="983" alt="image" src="https://user-images.githubusercontent.com/12693088/209006998-49238710-b5f2-458c-ae08-b7ebce904851.png">

It should automatically create a release with the changelog based on these rules via `.github/workflows/changelog.yml` when a tag is pushed.

All that needs to happen is triggering manual github action with desired version number:
<img width="347" alt="image" src="https://user-images.githubusercontent.com/12693088/209007415-6af91be1-e4f7-43c9-b8b0-cbc0b2b7b65e.png">

It can be improved where one could have options (major, minor, patch) instead of actual version, but that can be added later, no external packages necessary, just a few lines of ruby code.

arturictus commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,38 @@
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          ssh-key: ${{ secrets.RELEASE_SSH_KEY }}
```
do I need to set anything here?

andrcuns replied:
Just the ssh key as a secret named `RELEASE_SSH_KEY` the way I described in the above comment

andrcuns commented:
> Hey @andrcuns, this looks great! I added the secret: `GEM_HOST_API_KEY` What is the `RELEASE_SSH_KEY`? do we need it to publish the gem or it's for the release commit?

It is used for the manual release action in the step which automatically updates the version number, creates a tag and then pushed it back to repo:

```yml
-
        name: Update version
        run: |
          git config user.name github-actions
          git config user.email github-actions@github.com
          bundle config unset deployment
          bundle exec rake "version[${{ inputs.version }}]" && git push
      -
        name: Create tag and push to rubygems
        run: bundle exec rake release
        env:
          GEM_HOST_API_KEY: ${{ secrets.GEM_HOST_API_KEY }}
```

Basically it's ssh key that has access to push to repo.

It doesn't have to be your personal key. The way I did it, generated new key-pair and added public key as deploy key under security settings:
<img width="1129" alt="image" src="https://user-images.githubusercontent.com/12693088/209005305-c3be0f05-c5ea-4ff2-99c1-0b232ace24ce.png">
and added private key as secret:
<img width="1145" alt="image" src="https://user-images.githubusercontent.com/12693088/209005434-d0aa3e90-e47f-4c9d-8878-39359124233b.png">

It's not strictly necessary but it's nice to be able to have a full new version release cycle implemented via github actions

arturictus reviewed (commented):
looks amazing!!, almost there

arturictus commented:
> > Hey @andrcuns, this looks great! I added the secret: `GEM_HOST_API_KEY` What is the `RELEASE_SSH_KEY`? do we need it to publish the gem or it's for the release commit?
> 
> It is used for the manual release action in the step which automatically updates the version number, creates a tag and then pushed it back to repo:
> 
> ```yaml
> -
>         name: Update version
>         run: |
>           git config user.name github-actions
>           git config user.email github-actions@github.com
>           bundle config unset deployment
>           bundle exec rake "version[${{ inputs.version }}]" && git push
>       -
>         name: Create tag and push to rubygems
>         run: bundle exec rake release
>         env:
>           GEM_HOST_API_KEY: ${{ secrets.GEM_HOST_API_KEY }}
> ```
> 
> Basically it's ssh key that has access to push to repo.
> 
> It doesn't have to be your personal key. The way I did it, generated new key-pair and added public key as deploy key under security settings: <img alt="image" width="1129" src="https://user-images.githubusercontent.com/12693088/209005305-c3be0f05-c5ea-4ff2-99c1-0b232ace24ce.png"> and added private key as secret: <img alt="image" width="1145" src="https://user-images.githubusercontent.com/12693088/209005434-d0aa3e90-e47f-4c9d-8878-39359124233b.png">
> 
> It's not strictly necessary but it's nice to be able to have a full new version release cycle implemented via github actions

Done and Done

arturictus commented:
Added the secrets.
@andrcuns all ready to press merge. do you wanna do the honors?

andrcuns commented:
> Added the secrets. @andrcuns all ready to press merge. do you wanna do the honors?

Awesome, let's do it 🚀

andrcuns commented:
@arturictus What do You think about testing the release flow and releasing a new version?

It's a pretty substantial change, while there should be no breaking changes, not sure if it should be a minor or major release. Probably still minor, since it basically just expands support and refactors some things.

arturictus commented:
@andrcuns I think we can do a minor, let's try the releases 🤞

benlangfeld commented:
https://github.com/arturictus/sidekiq_alive/pull/80 fixes a bug that is present in the Sidekiq 7 compatibility here.
